Plugins/NativeXr/Source/NativeXr.cpp

```diff
@@ -1217,6 +1217,104 @@ namespace Babylon
             Napi::Value CreateAnchor(const Napi::CallbackInfo& info);
         };
 
+        // Implementation of the XRPlane interface: https://github.com/immersive-web/real-world-geometry/blob/master/plane-detection-explainer.md
+        class XRPlane : public Napi::ObjectWrap<XRPlane>
+        {
+            static constexpr auto JS_CLASS_NAME = "XRPlane";
+
+        public:
+            static void Initialize(Napi::Env env)
+            {
+                Napi::HandleScope scope{env};
+
+                Napi::Function func = DefineClass(
+                    env,
+                    JS_CLASS_NAME,
+                    {
+                        InstanceAccessor("planeSpace", &XRPlane::GetPlaneSpace, nullptr),
+                        InstanceAccessor("polygon", &XRPlane::GetPolygon, nullptr),
+                        InstanceAccessor("lastChangedTime", &XRPlane::GetLastChangedTime, nullptr),
+                    });
+
+                env.Global().Set(JS_CLASS_NAME, func);
+            }
+
+            static Napi::Object New(const Napi::Env& env)
+            {
+                return env.Global().Get(JS_CLASS_NAME).As<Napi::Function>().New({});
+            }
+
+            XRPlane(const Napi::CallbackInfo& info)
+                : Napi::ObjectWrap<XRPlane>{info}
+            {
+            }
+
+            void SetLastUpdatedTime(uint32_t timestamp)
+            {
+                m_lastUpdatedTimestamp = timestamp;
+            }
+
+            void SetNativePlane(xr::Plane& plane)
+            {
+                m_nativePlane = plane;
+            }
+
+            xr::Plane* GetNativePlane()
+            {
+                return &m_nativePlane;
+            }
+
+        private:
+            Napi::Value GetPlaneSpace(const Napi::CallbackInfo& info)
+            {
+                Napi::Object napiTransform = XRRigidTransform::New(info);
+                XRRigidTransform* rigidTransform = XRRigidTransform::Unwrap(napiTransform);
+                rigidTransform->Update(m_nativePlane.Center);
+
+                Napi::Object napiSpace = XRReferenceSpace::New(info.Env(), napiTransform);
+                return std::move(napiSpace);
+            }
+
+            Napi::Value GetPolygon(const Napi::CallbackInfo& info)
+            {
+                // Translate the polygon from a native array to a JS array.
+                auto polygonArray = Napi::Array::New(info.Env(), m_nativePlane.PolygonSize);
+                for (size_t i = 0; i < m_nativePlane.PolygonSize; i++)
+                {
+                    auto polygonPoint = Napi::Object::New(info.Env());
+                    if (m_nativePlane.PolygonFormat == xr::PolygonFormat::XZ)
+                    {
+                        size_t polygonIndex = 2 * i;
+                        polygonPoint.Set("x", m_nativePlane.Polygon[polygonIndex]);
+                        polygonPoint.Set("y", 0);
```

Comment:
I wonder if it would be simpler to just always have XYZ and make it an implementation detail of ARCore that the y value is always 0 (that the planar polygon is always on the XZ plane)?

Reply:
We get the polygon buffer directly from ARCore, which is formatted as 2D points.  I think it's better to leave it as is than have to copy it again.